### DISCUSSION TITLE
Smaller granularity of access control level based on namespace, like java's default?

### DISCUSSION THREAD

**Discussion by acaly** [dotnet/csharplang#4092] [Language Ideas] [answered]
I have seen discussions in this repo regarding friend class. Sometimes I also face similar situation and feel that the access control is a little bit limited in C#. I am not sure whether a similar feature to Java's default access level (visible to classes in the same package, counterpart for namespace in C#) have been considered.

**What's it like:**
A new keyword that specifies a new access level for type members, which can only be accessed by other types which are: within the same assembly (and other InternalsVisibleTo assemblies) AND within the same namespace. Access from sub-namespace might or might not be allowed. 

**How it might be implemented:**
As have bee pointed out in the friend class topic, it can be emitted as an internal member with a special attribute. Maybe an [Obsolete] can also be added so that older compilers can at least generate warnings. IDE can filter these members when doing like auto-completion, and C# compiler can generate errors if the rule is not obeyed. This avoids adding new runtime features.

**Why I think it's good:**
Currently the only tool for us is internal (plus InternalsVisibleTo) and nested types. Internal is not very selective, and nested types is also limited (you can't put A inside B while also put B inside A, etc.). A namespace based granularity would allow us to put related classes (friend classes) together in a separate namespace and share their members selectively. I hope it solves most issues those who wants the friend class features have met.

Many people do not prefer friend class probably because:
* A friend class is a special one-way relationship. This requires adding some additional information to each relationship, which seems a little bit dirty, and we currently don't have similar special relations from type to type in CLR.
* It is not clear whether friend relationship can be inherited (on both sides), or propagated (although I believe most will think no).
* People keep secrets even to friends, so exposing all private members is not a good choice.

I am curious how you think about this. Thanks!

**Comment by svick** [comment, marked as answer]
This is pretty much a duplicate of https://github.com/dotnet/csharplang/discussions/1138.

**Reply by CyrusNajmabadi** [reply]
This can also be implemented pretty trivially with an analyzer. A good rule of thumb is: of you're trying to restrict normal code so that it is now illegal, just used an analyzer to do that first.

**Reply by acaly** [reply]
Thank you for the information. I didn't see that thread.

Using analyzer sounds like a good and clean solution. I will try it.
